samples/nrf9160/cloud_simple/src/main.c

```diff
@@ -0,0 +1,249 @@
+	if (err) {
+		printk("cloud_send failed, error: %d\n", err);
+	}
+
+	k_delayed_work_submit(
+			&cloud_update_routine_work,
+			K_SECONDS(CONFIG_CLOUD_MESSAGE_PUBLICATION_INTERVAL));
+}
+
+void cloud_event_handler(const struct cloud_backend *const backend,
+			 const struct cloud_event *const evt,
+			 void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	switch (evt->type) {
+	case CLOUD_EVT_CONNECTED:
+		printk("CLOUD_EVT_CONNECTED\n");
+		break;
+	case CLOUD_EVT_READY:
+		printk("CLOUD_EVT_READY\n");
+		k_delayed_work_submit(&cloud_update_routine_work, K_NO_WAIT);
+		break;
+	case CLOUD_EVT_DISCONNECTED:
+		printk("CLOUD_EVT_DISCONNECTED\n");
+		break;
+	case CLOUD_EVT_ERROR:
+		printk("CLOUD_EVT_ERROR\n");
+		break;
+	case CLOUD_EVT_DATA_SENT:
+		printk("CLOUD_EVT_DATA_SENT\n");
+		break;
+	case CLOUD_EVT_DATA_RECEIVED:
+		printk("CLOUD_EVT_DATA_RECEIVED\n");
+		printk("Data received from cloud: %s\n", evt->data.msg.buf);
+		break;
+	case CLOUD_EVT_PAIR_REQUEST:
+		printk("CLOUD_EVT_PAIR_REQUEST\n");
+		break;
+	case CLOUD_EVT_PAIR_DONE:
+		printk("CLOUD_EVT_PAIR_DONE\n");
+		break;
+	case CLOUD_EVT_FOTA_DONE:
+		printk("CLOUD_EVT_FOTA_DONE\n");
+		break;
+	default:
+		printk("Unknown cloud event type: %d\n", evt->type);
+		break;
+	}
+}
+
+/**@brief Initialize delayed work items. */
+static void work_init(void)
+{
+	k_delayed_work_init(&cloud_update_routine_work,
+			    cloud_update_routine_work_fn);
+}
+
+/**@brief Configures modem to provide LTE link. Blocks until link is
+ * successfully established.
+ */
+static void modem_configure(void)
+{
+#if defined(CONFIG_BSD_LIBRARY)
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already turned on
+		 * and connected.
+		 */
+	} else {
+		int err;
+
+		printk("Connecting to LTE network. ");
+		printk("This may take several minutes.\n");
+
+		err = lte_lc_init_and_connect();
+		if (err) {
+			printk("LTE link could not be established.\n");
+			error_handler(ERROR_LTE_LC, err);
+		}
+
+		printk("Connected to LTE network\n");
+	}
+#endif
+}
+
+void main(void)
+{
+	int err;
+
+	printk("Cloud simple has started\n");
+
+	cloud_backend = cloud_get_binding("NRF_CLOUD");
+	__ASSERT(cloud_backend != NULL, "NRF_CLOUD backend not found");
+
+	err = cloud_init(cloud_backend, cloud_event_handler);
+	if (err) {
+		printk("Cloud backend could not be initialized, error: %d\n",
+			err);
+		error_handler(ERROR_CLOUD, err);
+	}
+
+	work_init();
+	modem_configure();
+
+	err = cloud_connect(cloud_backend);
+	if (err) {
+		printk("cloud_connect failed: %d\n", err);
+		error_handler(ERROR_CLOUD, err);
+	}
+
+	struct pollfd fds[] = {
+		{
+			.fd = cloud_backend->config->socket,
+			.events = POLLIN
+		}
+	};
+
+	while (true) {
+		/* The timeout is set to (keepalive / 3), so that the worst case
+		 * time between two messages from device to broker is
+		 * ((4 / 3) * keepalive + connection overhead), which is within
+		 * MQTT specification of (1.5 * keepalive) before the broker
```

Comment:
I'm not fully happy having the MQTT being mentioned here, while it's a generic sample. While I understand that both existing cloud backends are MQTT-based, this indicates that we should either hide the poll functionality underneath cloud backend (sacrificing possibility to poll on other sockets as well) or provide some helper function to retrieve sleep time (as we've briefly discussed in regards of MQTT @jtguggedal).

But it's more of a improvement for the future, I guess we can live with that for now.

Reply:
Agree that it should be fixed somehow in the (near) future. A suggestion is to use `mqtt_ping()` instead of `mqtt_live()` in the cloud backends to ensure we don't go paste the keep-alive deadline. There will be some unnecessary pings sent, but at least we will be within spec without workarounds.

Reply:
And also, I think we should add Kconfig option to run `poll` in a thread in the backend. As @rlubos says, user will lose ability to poll other sockets, and we'll have to add error events for poll errors.